File: python/periph/transport/neopixel_linux.py
import spidev

from .base import Transport


def _encode(data):
    """Encode NeoPixel data bytes to SPI bitstream.

    Each NeoPixel bit is encoded as 3 SPI bits:
    - 0: 0b100 (417ns high, 833ns low)
    - 1: 0b110 (833ns high, 417ns low)
    """
    out = bytearray(len(data) * 3 + 16)
    for i, byte in enumerate(data):
        bits = 0
        for bit in range(7, -1, -1):
            bits = (bits << 3) | (0b110 if (byte >> bit) & 1 else 0b100)
        out[i * 3:(i + 1) * 3] = bits.to_bytes(3, 'big')
    return bytes(out)


class NeoPixelTransport(Transport):
    """NeoPixel transport using spidev.

    Drives WS2812B-compatible addressable LEDs over a single data line using
    timing-encoded NZR protocol via SPI MOSI at 800 Kbps.

    The caller is responsible for color ordering and bytes-per-pixel.
    """

    def __init__(self, bus_num, device_num):
        """Construct a NeoPixel transport.

        Args:
            bus_num: SPI bus number (e.g. 0 for /dev/spidev0.0).
            device_num: SPI device number (e.g. 0 for /dev/spidev0.0).
        """
        self._spi = spidev.SpiDev()
        self._spi.open(bus_num, device_num)
        self._spi.mode = 0
        self._spi.max_speed_hz = 2_400_000

    def write(self, data):
        """Encode and transmit NeoPixel data.

        Args:
            data: bytes of pixel data (3 bytes/pixel for RGB, 4 for RGBW).
        """
        encoded = _encode(data)
        self._spi.xfer2(list(encoded))

    def close(self):
        """Release the SPI device."""
        self._spi.close()

File: go/periph/transport/neopixel_linux.go
//go:build linux && !tinygo

// NeoPixelTransport is the Linux implementation of the NeoPixel write-only
// transport. It wraps the Go Linux SPI transport opened at 2.4 MHz, mode 0,
// and encodes each NeoPixel bit as three SPI bits to produce the timing
// that WS2812B / SK6812 LED strips expect.
package transport

// NeoPixelTransport is the Linux /dev/spidevB.D-backed implementation
// of the NeoPixel write-only transport. It holds an SPITransport opened
// at 2.4 MHz, mode 0, MSB-first, and encodes each NeoPixel bit as
// three SPI bits: `100` for a 0 bit, `110` for a 1 bit. The
// resulting buffer is sent in a single SPI transfer followed by 16
// zero bytes for the reset.
type NeoPixelTransport struct {
	spi *SPITransport
}

// NewNeoPixelTransport opens /dev/spidevBUS.DEVICE at 2.4 MHz, mode 0,
// for NeoPixel bit-encoding. No CS pin is required for WS2812B/SK6812
// strips (CS is a dummy in the kernel's view but still toggled, which
// the strips ignore).
func NewNeoPixelTransport(busNum, deviceNum int) (*NeoPixelTransport, error) {
	spi, err := NewSPITransport(busNum, deviceNum, 2_400_000)
	if err != nil {
		return nil, err
	}
	return &NeoPixelTransport{spi: spi}, nil
}

// Close releases the underlying SPI device.
func (n *NeoPixelTransport) Close() error {
	return n.spi.Close()
}

// Write encodes the buffer and transmits it to the LED strip in a
// single SPI transfer. Each NeoPixel byte is encoded as three SPI
// bytes; 16 trailing zero bytes produce the reset pulse.
func (n *NeoPixelTransport) Write(data []byte) error {
	encoded := encodeNeoPixel(data)
	return n.spi.Write(encoded)
}

// encodeNeoPixel converts a buffer of NeoPixel bytes into 3-bits-per-bit
// SPI bytes, with 16 trailing zero bytes for the reset. Each NeoPixel
// bit maps to 3 SPI bits at 2.4 MHz: `100` for a 0, `110` for a 1.
func encodeNeoPixel(data []byte) []byte {
	out := make([]byte, len(data)*3+16)
	for i, b := range data {
		var bits uint32
		for bit := 7; bit >= 0; bit-- {
			pattern := uint32(0b100)
			if (b>>bit)&1 == 1 {
				pattern = 0b110
			}
			bits = (bits << 3) | pattern
		}
		out[i*3+0] = byte(bits >> 16)
		out[i*3+1] = byte(bits >> 8)
		out[i*3+2] = byte(bits)
	}
	return out
}

// Read is a no-op for NeoPixel: strips are write-only. Provided to
// satisfy the Transport interface.
func (n *NeoPixelTransport) Read(count int) ([]byte, error) {
	return nil, nil
}

// WriteRead is a no-op for NeoPixel: strips are write-only. Provided
// to satisfy the Transport interface.
func (n *NeoPixelTransport) WriteRead(data []byte, count int) ([]byte, error) {
	return nil, nil
}
